### ISSUE TITLE
Missing error when mistyped value overwrites property of spreaded generic type parameter

### ISSUE THREAD

**Issue by just-boris** [microsoft/TypeScript#42690]
# Bug Report

### 🔎 Search Terms

generic argument, rest spread, false positive

### 🕗 Version & Regression Information

- [x] This is the behavior in every version I tried, and I reviewed the FAQ for entries about object spread and generics

I found a few relevant issues, but none of them fully correspond to the current one:
* https://github.com/microsoft/TypeScript/issues/35858
* https://github.com/microsoft/TypeScript/issues/38469
* https://github.com/microsoft/TypeScript/issues/30129

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-dev.20210207#code/JYOwLgpgTgZghgYwgAgJIgA4FcwAUoD2GAzsgN4CwAUMsgG5wA2WEAXMsWFKAOYDc1AL7VqCAiE7IeEPIRLIAvMgAUGOcXbpssosQCUmzDny7FAPhWUayAHR21ugDTVaDZm2QBGAEwBmR8gA9IHIAO4AFhAgyGCRUCjApCAEUlHQwAgxAJ4YEAHAYBzhBKGkcNHQhFABcKQQAB65CJAAJkJ6AlSi4pLSOiQA4mncCAByBADKuXAtisgAPAAqyA2QIC2kWsbqZqrq7IsGyMsKFla0YhKFIBChJvJKZHY2DiSCnbQ3d+o2bixzPl8fCCISYxBS0AI1WQtRWjQgzQgbWs8TAWCg0S+92InXeIi6VEuvRk2KGNxGcyWcLWGzQRn6xF2rw0xyOJwsynOtns6mc1j+HkBAWCyGSKygVWQcQg7U61CAA)

### 💻 Code

```ts
interface InputProps {
  value: string;
}

const getProps = (props: InputProps): InputProps => ({
  ...props,
  value: 123, // when there is no generic type, it shows an error, as expected
});

const getPropsGenericNoSpead = <T extends InputProps>(props: T): T => {
  const newProps = {...props};
  newProps.value = 123; // also error, as expected
  return newProps;
};


const getPropsGeneric = <T extends InputProps>(props: T): T => ({
  ...props,
  value: 123, // no error here
});
```

### 🙁 Actual behavior

Typescript allows assigning incompatible types

### 🙂 Expected behavior

Typescript shows an error

**Comment by misuken-now** [comment]
I often run into the same problem and have trouble with it.

Patterns encountered in React.

[playground](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=14&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwCwAUHI1gB6Sxy4QB2q8AYRKQeWHjABcmKjAB0ANQBiAgDwBvOMAAmAfin8owHgHM4AXwB8cALxwVAFThsYYreg3a9cA0dOWAFGDEYKhS9gCUNlZqjHB2FrFxcACMsnZawABucGqyeUEQIWZwAPRWSXGAAAyAoORwAPIA0olxAExpKhnZ2tZqycVlFYNJtQCiUMRQzXAAzO2dOXmyBUWaWj19pVaAfBuAKLv1DcxMdmX0TIncfPBQyWE2OXCLy6gANKtSyeY0pSVwu-uxiUoMAArlAeHAeMCADZQ05mRhAA)
```tsx
import * as React from "react";
 
export const Component: React.VFC<{ id?: string }> = <T extends { id?: string }>(props: T) => {
  <>
    1. <div {...props} />      　✅ OK
    2. <div id={1} />            ✅ Error
    3. <div {...props} id={1} /> 🤔 OK 
  </>;
 
  const r1: T = { ...props, id: 1 }; // 🤔 OK
 
  return null;
}
```

**Comment by misuken-now** [comment]
I was able to write a function to check the problem.

```ts
function f1<T extends { id: string }, U extends { id: number }>(props: T, props2: U) {
    function valid1<V, U>(props: U & {[P in keyof V]?: V[P]}) { return props; };
    function valid2<V, U>(props: U & {[P in keyof V]?: V[P]}) { return props as U; };
    const R11 = valid1({ ...props, ...props2 }); // never ✅ Argument Error
    const R12 = valid1({ ...props, id: 1 });     // never ✅ Argument Error
    const R13 = valid1({ ...props, id: "" });    // T & { id: string; } & { id?: string | undefined; } ✅ OK

    const R21 = valid2({ ...props, ...props2 }); // T & U
    const R22 = valid2({ ...props, id: 1 });     // T & { id: number; } ✅ Argument Error
    const R23 = valid2({ ...props, id: "" });    // T & { id: string; } ✅ OK
}
```

[playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMAjAHgCqIKYA8rZgAmAzogN6IxEBciJUATjGAOaIC+ANIgKo75CpClVqIwIALYAjbI04A+ABQAHRnBUk6GHmo0kATHV4BKCgChEV5OGjwkEABbYIAawwBPFdnQA1HrzKeprGiABkFADaAApUSK7YHnDAiL4AugD8dL4xaRxmlIzYUCCMSMEkANyclZbWoJCwCIhOLu5e2AZo-nxB6iF84VGxLIgJSSnpWam5+SJFJWWIFYgAhmS81Ry11i0IDIgASigoiAC8Lc5unt4oSpQAdE8VPE8PFQacJtUA9D-i2AAbnJEIBQckQAEFGKwpIQoIgAKKMdSMOpWCD7eHHT4XVrXDp3R7PfokHjUOinfLVXZ-AHA+TgqEwyRwxHIuCo3YYsAHY4AZnOlzaNx890QbxeojoACJpV9qVZaVgIpRyfQmCxWFshqqiNMGMw2IgAD6IcBEbDAFjYIja8EAeQA0uY0XseViDKdcVd2t4DGKJSTXsT9J8qYglUNeK7ubyDDihfi-QGQ5oyWJKd9dhH-sqRGqJDI5HbIdDYWB4UiUTHMUcDALvcKOv6ie8g1LELL5TTczqOwbNSWneYOOYgA)

**Comment by vkarponen** [comment]
It would seem that the problem appears also without generics:
```
function test(param: { [key: string]: (string | number) }) {
    const stringValues: { [key: string]: string } = {
        foo: 'bar',
        ...param
    };
    console.log(stringValues);
}
```
is falsely not considered an error but if you remove `foo: 'bar',` from the above then TS considers that to be an error as it should.

**Comment by misuken-now** [comment]
In situations where generics are not used, the use of "Object.assign" may be a solution.

```ts
function test(param: { [key: string]: (string | number) }) {
  // Type '{ foo: string; } & { [key: string]: string | number; }' is not assignable to type '{ [key: string]: string; }'.
  // 'string' index signatures are incompatible.
  //   Type 'string | number' is not assignable to type 'string'.
  //     Type 'number' is not assignable to type 'string'.(2322)
  const stringValues: { [key: string]: string } = Object.assign({}, {
    foo: 'bar'
  }, param);
  console.log(stringValues);
}
```

[playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFApgZygCgA4EMBOuAtgFyIDeiA2gNYoCeZG+MYA5gLpmbOtuIAfRGBBEARinwBKRAF8Z5AFCJEAelWIAKvWwpEAckrA4cJlBbsA3HMQAyCtTqNEvdlxfm+g4aIn5rsvqIMGjCcFCIuGhoMGxguGIANnpQcMg6eoaODGYWnLl8AfoAdMpqGvqubEGsACYoAB4usfFQIPjokR3BkHBEeLBJKKUq6ipaGQZV3iLikjWhYOGR0S0JychpUJOVnuwlZWPjE7oGs34LYRFRMXHrKVs7VSWYAEwAzK+vUmUQCBgePIANVwiRA6DIlFoOUBfHc01kiAAvIgAPJiABWKGgxRuLUw5FkABoKGUVMZTAYxAR9GViYg8IQiFJLL9-nBksVEnA2Dw9mwQWD0CzFLIgA)

**Comment by teynar** [comment]
I am also experiencing this issue and I feel quite surprised since this looks like a fairly common case when using reduce with the spread operator.

```ts
function foo<T extends string>(keys: T[]): Record<T, number> {
  return keys.reduce((acc, key) => {
    return {
      ...acc,
      [key]: 'string' // this is not a number, but there is no error
    }
  }, {} as Record<T, number>);
}
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdESoAnGMAcwD4AKAa2wE8SAuRDAbQF0AlNwBK2CHEpFMAGkRgQAWwBG2SvUQBvAFCJElbFBCUkrDgDoDREBGyNGAQwgQ5ZwYgC8GnXr0GjJlq6vnrmYU4uwSG8ZvzcAOQAjPGIAPSpiFAAFjBkufJwUIgO8kqqlHLKIEXZatiI+WBwOJSUklGIAL7BnXKancVkYhJSsqUqavSCANzanUA)

**Comment by kumikumi** [comment]
This is probably related to the same issue:
